// @jeefo/publish/package_version_manager.js
/* -.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.
* File Name   : package_version_manager.js
* Created at  : 2019-10-31
* Updated at  : 2019-10-31
* Author      : jeefo
* Purpose     :
* Description :
* Reference   :
.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.*/
// ignore:start
"use strict";

/* globals*/
/* exported*/

// ignore:end

const fs             = require("fs").promises;
const PackageVersion = require("./package_version");

const save_json = async (filepath, json) => {
    const data = JSON.stringify(json, null, 4);
    await fs.writeFile(filepath, data, "utf8");
};

const update_package_json = async ({ pkg, version, package_path }) => {
    pkg.version = version.toString();
    await save_json(package_path, pkg);
};

class PackageVersionManager {
    constructor (working_directory, saving_directory = working_directory) {
        this.pkg          = null;
        this.version      = null;
        this.package_path = `${ saving_directory }/package.json`;

        try {
            this.pkg     = require(`${ working_directory }/package.json`);
            this.version = new PackageVersion(this.pkg.version);
        } catch (e) {
            console.error("Cannot read package.json");
            process.exit();
        }
    }

    increment_major () {
        this.version.major += 1;
        update_package_json(this);
    }
    increment_middle () {
        this.version.middle += 1;
        update_package_json(this);
    }
    increment_minor () {
        this.version.minor += 1;
        update_package_json(this);
    }
}

module.exports = PackageVersionManager;
